multishard_mutation_query.cc

```diff
@@ -508,6 +504,23 @@ future<> read_context::save_reader(shard_id shard, const dht::decorated_key& las
                     reader->unpop_mutation_fragment(mutation_fragment_v2(schema, rparts->permit, *rit));
                 }
             }
+            if (rm.compaction_state) {
+                auto* next_fragment = co_await reader->peek();
+                // If there is nothing more left from the current partition --
+                // next fragment is either partition start, partition end or
+                // EOS -- then we don't push the partition header back to the
+                // reader.
+                if (next_fragment && (!next_fragment->is_partition_start() && !next_fragment->is_end_of_partition())) {
+                    auto& cs = *rm.compaction_state;
+                    if (cs.current_tombstone) {
+                        reader->unpop_mutation_fragment(mutation_fragment_v2(schema, rparts->permit, range_tombstone_change(*cs.current_tombstone)));
+                    }
+                    if (cs.static_row) {
+                        reader->unpop_mutation_fragment(mutation_fragment_v2(schema, rparts->permit, static_row(schema, *cs.static_row)));
+                    }
+                    reader->unpop_mutation_fragment(mutation_fragment_v2(schema, rparts->permit, partition_start(cs.partition_start)));
+                }
+            }
```

Comment:
Please extract, with testcase.

Reply:
I'm not sure I will be able to write a test-case for this. In this series I have a test-case that in theory should reproduce it 100% of the time, yet it reproduces only part of the time. At one point the same test would reproduce it on my laptop, but on my desktop. This suggests it has a timing component though I don't understand why.

Reply:
It's worth to spend some time to understand it, if we have some rare bug lurking somewhere.

Reply:
https://github.com/scylladb/scylla/pull/11137
Reproducing the bug doesn't depend on timing after all: in the context of the test I wrote for this PR it depends on the distribution of partitions across vnodes. I used a random key generating method to generate my partition keys, I think it was just a matter of lucky/unlucky seeds.
With a targeted test-case the issue is reproducible 100% of the time.

Reply:
I'm glad I made you chase this even though it probably wasn't fun, I think it's an important bug.